File: tests/test-terraria/src/main/java/pize/tests/terraria/map/WorldMap.java
package pize.tests.terraria.map;

import static pize.tests.terraria.tile.TileType.AIR;

public class WorldMap{

    private final int width, height;
    private final MapTile[] tiles;

    public WorldMap(int width, int height){
        this.width = width;
        this.height = height;
        tiles = new MapTile[width * height];

        for(int i = 0; i < tiles.length; i++)
            tiles[i] = new MapTile(AIR, 15);
    }


    public MapTile getTile(int x, int y){
        if(isOutOfBounds(x, y))
            return null;

        return tiles[getIndex(x, y)];
    }


    public int getWidth(){
        return width;
    }

    public int getHeight(){
        return height;
    }


    public MapTile[] getTiles(){
        return tiles;
    }

    public int getIndex(int x, int y){
        return y * width + x;
    }

    public boolean isOutOfBounds(int x, int y){
        return x >= width || y >= height || x < 0 || y < 0;
    }

}

File: tests/test-terraria/src/main/kotlin/pize/tests/terraria/map/WorldMap.kt
package pize.tests.terraria.map

import pize.tests.terraria.tile.TileType

class WorldMap(val width: Int, val height: Int) {
    val tiles: Array<MapTile?>

    init {
        tiles = arrayOfNulls(width * height)
        for (i in tiles.indices) tiles[i] = MapTile(TileType.AIR, 15)
    }

    fun getTile(x: Int, y: Int): MapTile? {
        return if (isOutOfBounds(x, y)) null else tiles[getIndex(x, y)]
    }

    fun getIndex(x: Int, y: Int): Int {
        return y * width + x
    }

    fun isOutOfBounds(x: Int, y: Int): Boolean {
        return x >= width || y >= height || x < 0 || y < 0
    }
}